lib/seek/merging/person_merge.rb

```diff
@@ -0,0 +1,101 @@
+
+      private
+
+      def merge_simple_attributes(other_person)
+        # This attributes are directly copied from other_person if they are empty in the person to which its merging.
+        # first_letter is also updated
+        simple_attributes = %i[first_name last_name email phone skype_name web_page description avatar_id orcid]
+        simple_attributes.each do |attribute|
+          send("#{attribute}=", other_person.send(attribute)) if send(attribute).nil?
+        end
+        update_first_letter
+      end
+
+      def merge_annotations(other_person)
+        annotation_types = %w[expertise tools]
+        annotation_types.each do |annotation_type|
+          add_annotations(send(annotation_type) + other_person.send(annotation_type), annotation_type.singularize, self)
+        end
+
+        # Update annotations by other_user
+        return unless user && other_person.user
+
+        merge_user_associations(other_person, 'annotations_by',
+                                %i[annotatable_type annotatable_id value_id],
+                                { source_id: user.id })
+      end
+
+      def merge_resources(other_person)
+        # Contributed
+        Person::RELATED_RESOURCE_TYPES.each do |resource_type|
+          other_person.send("contributed_#{resource_type.underscore.pluralize}").in_batches.update_all(contributor_id: id)
+        end
+        # Created
+        duplicated = other_person.created_items.pluck(:id) & created_items.pluck(:id)
+        AssetsCreator.where(creator_id: other_person.id, asset_id: duplicated).in_batches.destroy_all
+        AssetsCreator.where(creator_id: other_person.id).in_batches.update_all(creator_id: id)
+      end
+
+      def merge_associations(other_person, assoc, duplicates_match, update_hash)
+        other_items = other_person.send(assoc).pluck(*duplicates_match)
+        self_items = send(assoc).pluck(*duplicates_match)
+        duplicated = other_items & self_items
+        duplicated = duplicated.map { |item| [item] } if duplicates_match.length == 1
+
+        duplicated_hash = Hash[duplicates_match.zip(duplicated.transpose)]
+        other_person.send(assoc).where(duplicated_hash).in_batches.destroy_all
+
+        other_person.send(assoc).in_batches.update_all(update_hash)
+      end
+
+      def merge_user(other_person)
+        return unless user && other_person.user
+
+        merge_user_associations(other_person, 'identities',
+                                %i[provider uid], { user_id: user.id })
+        merge_user_associations(other_person, 'oauth_applications',
+                                %i[redirect_uri scopes], { owner_id: user.id })
+        merge_user_associations(other_person, 'access_tokens',
+                                %i[application_id scopes], { resource_owner_id: user.id })
+        merge_user_associations(other_person, 'access_grants',
+                                %i[application_id redirect_uri scopes], { resource_owner_id: user.id })
+      end
```

Comment:
I did a search for `belongs_to :user` and found a few other associations:
![image](https://github.com/seek4science/seek/assets/503373/2e542b2a-ac89-4417-bf7b-f00cf1ad6dc4)

The auth lookup tables will be too slow to update, but the various log tables are probably worth updating

Reply:
the auth lookup table should update automatically. rows will be removed when the user is destroyed (another case where we can enhance the delete_all). After it's merged I also plan to check what jobs get created, or when it just needs a job firing at the end